dtm0/fop.c

```diff
@@ -0,0 +1,405 @@
+
+	req               = m0_fop_data(fop);
+	req->dtr_msg      = msg_type;
+	rc = m0_dtm0_tx_desc_copy(txd, &req->dtr_txr) ?: m0_rpc_post(item);
+	if (rc == 0) {
+		phase_sm_id = m0_sm_id_get(&fom->fo_sm_phase);
+		rpc_sm_id   = m0_sm_id_get(&item->ri_sm);
+		M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, phase_sm_id, rpc_sm_id);
+
+		m0_fop_put_lock(fop);
+
+		M0_LOG(M0_DEBUG, "Sent %d msg " FID_F " -> " FID_F, msg_type,
+		       FID_P(&dtms->dos_generic.rs_service_fid), FID_P(tgt));
+	}
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int m0_dtm0_logrec_update(struct m0_be_dtm0_log  *log,
+				      struct m0_be_tx        *tx,
+				      struct m0_dtm0_tx_desc *txd,
+				      struct m0_buf          *pyld)
+{
+	int rc;
+
+	M0_ENTRY();
+
+	m0_mutex_lock(&log->dl_lock);
+	rc = m0_be_dtm0_log_update(log, tx, txd, pyld);
+	m0_mutex_unlock(&log->dl_lock);
+
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom                *fom,
+				     const struct m0_dtm0_tx_desc *txd)
+{
+	struct m0_dtm0_service *dtms = m0_dtm0_service_find(
+		fom->fo_service->rs_reqh);
+	struct m0_dtm0_tx_desc  msg = {};
+	int                     rc;
+	int                     i;
+
+	rc = m0_dtm0_tx_desc_copy(txd, &msg);
+	if (rc != 0)
+		return M0_ERR(rc);
+
+	/*
+	 * TODO: This change will be done by DTM0 log because
+	 * it should log the entry "pa == self" with PERSISTENT state
+	 * set.
+	 */
+	for (i = 0; i < msg.dtd_ps.dtp_nr; ++i) {
+		if (m0_fid_eq(&msg.dtd_ps.dtp_pa[i].p_fid,
+			      &dtms->dos_generic.rs_service_fid)) {
+			msg.dtd_ps.dtp_pa[i].p_state =
+				max_check(msg.dtd_ps.dtp_pa[i].p_state,
+					  (uint32_t) M0_DTPS_PERSISTENT);
+		}
+	}
+
+	/* Notify the originator */
+	rc = m0_dtm0_send_msg(fom, DTM_PERSISTENT, &msg.dtd_id.dti_fid, &msg);
+
+	/* TODO: Send P msgs to the rest of the participants. */
+	m0_dtm0_tx_desc_fini(&msg);
+
+	return M0_RC(rc);
+}
+
+
+static int dtm0_fom_tick(struct m0_fom *fom)
+{
+	int                     rc;
+	struct dtm0_req_fop    *req;
+	struct dtm0_rep_fop    *rep;
+	struct m0_dtm0_service *svc;
+
+	if (m0_fom_phase(fom) < M0_FOPH_NR) {
+		rc = m0_fom_tick_generic(fom);
+	} else {
+		M0_ASSERT(m0_fom_phase(fom) == M0_FOPH_TYPE_SPECIFIC);
+		req = m0_fop_data(fom->fo_fop);
+		rep = m0_fop_data(fom->fo_rep_fop);
+		M0_SET0(&rep->dr_txr);
+		svc = m0_dtm0_service_find(fom->fo_service->rs_reqh);
+		M0_ASSERT(svc != NULL);
+
+		M0_LOG(M0_DEBUG, "Processing non-generic phase %" PRIu32
+		       " %d, " FID_F ", %p", req->dtr_msg,
+		       !!m0_dtm0_is_a_volatile_dtm(fom->fo_service),
+		       FID_P(&fom->fo_service->rs_service_fid),
+		       fom->fo_service->rs_reqh);
+
+		if (m0_dtm0_in_ut() && req->dtr_msg == DMT_EXECUTE &&
+		    m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
+			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr,
+						  &rep->dr_txr);
+			M0_ASSERT(rc == 0);
+			rc = m0_dtm0_send_msg(fom, DMT_EXECUTED,
+					      &req->dtr_txr.dtd_id.dti_fid,
+					      &req->dtr_txr);
+			M0_ASSERT(rc == 0);
```

Comment:
error handling

Reply:
Will be handled during working on failure path.